File: backend/models/course.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class CourseCreate(BaseModel):
    university_name: str
    course_name: str
    course_code: Optional[str] = None
    professor_name: Optional[str] = None


class GradingCategory(BaseModel):
    name: str
    weight: float  # 0.0 to 1.0
    notes: Optional[str] = None


class DetectedTool(BaseModel):
    tool_name: str
    evidence: str
    purpose: str
    confidence: float
    integration_type: str  # api | extension_scrape | manual


class PublicResource(BaseModel):
    title: str
    url: str
    type: str  # github_repo | notes | slides | assignments | other
    relevance_score: float
    reason: str


class StudentSignal(BaseModel):
    workload: Optional[str] = None  # low | medium | high
    difficulty: Optional[str] = None
    grading_style: Optional[str] = None
    key_warnings: list[str] = []
    positive_signals: list[str] = []
    summary: str = ""
    confidence: float = 0.0


# ── Response models ──────────────────────────────────────────────────────────

class WeeklyAction(BaseModel):
    title: str
    rationale: str
    priority: int
    due_date: Optional[str] = None


class ActionPlanResponse(BaseModel):
    risk_level: str
    risk_explanation: str = ""
    weekly_actions: list[WeeklyAction] = []
    missing_data_flags: list[str] = []
    focus_note: str = ""


class BootstrapResponse(BaseModel):
    id: Optional[str] = None
    course_identity: dict = {}
    syllabus_status: dict = {}
    course_profile: dict = {}
    resources: list = []
    detected_tools: list = []
    student_signal: dict = {}
    obligations: list = []

File: frontend/types/course.ts
export interface CourseIdentity {
  canonical_name: string;
  course_code: string;
  university: string;
  official_links: string[];
  confidence: number;
  notes: string;
}

export interface GradingCategory {
  name: string;
  weight: number; // 0.0 – 1.0
  notes?: string | null;
}

export interface DetectedTool {
  tool_name: string;
  evidence: string;
  purpose: string;
  confidence: number;
  integration_type: "api" | "extension_scrape" | "manual";
}

export interface PublicResource {
  title: string;
  url: string;
  type: "github_repo" | "notes" | "slides" | "assignments" | "other";
  relevance_score: number;
  reason: string;
}

export interface StudentSignal {
  workload?: string | null;
  difficulty?: string | null;
  grading_style?: string | null;
  key_warnings: string[];
  positive_signals: string[];
  summary: string;
  confidence: number;
}

export interface SyllabusStatus {
  found: boolean;
  confidence: number;
  source: "upload" | "web_search";
}

export interface CourseProfile {
  grading_categories: GradingCategory[];
  late_policy?: string;
  attendance_policy?: string;
  drop_rules?: string;
  required_tools?: string[];
  key_deadlines?: { title: string; date: string; type: string }[];
  course_summary?: string;
  workflow_notes?: string;
}

export interface BootstrapResponse {
  course_identity: CourseIdentity;
  syllabus_status: SyllabusStatus;
  course_profile: CourseProfile;
  resources: PublicResource[];
  detected_tools: DetectedTool[];
  student_signal: StudentSignal;
}

/** Local aggregate stored in Zustand; combines bootstrap data + metadata */
export interface Course {
  id: string;
  university: string;
  courseName: string;
  courseCode: string;
  professor: string;
  createdAt: string;
  bootstrap: BootstrapResponse;
}
